internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,178 @@
+package addonsdk
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	statusReporterSingleton *StatusReporter
+	initializeSingletonOnce sync.Once
+)
+
+// a nice heartbeat-reporter implemented by the MT-SRE which our tenants can use
+// if they don't like it, they can implement their own heartbeat reporter by creating a type which implements the `addonsdk.statusReporterClient` interface
+type StatusReporter struct {
+	// object provided by the client/tenants which implements the addonsdk.client interface
+	addonInstanceInteractor client
+	addonName               string
+	addonTargetNamespace    string
+
+	// the latest conditions which the heartbeat reporter would be reporting periodically
+	latestConditions []metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	interval time.Duration
+	ticker   *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	updateCh chan updateOptions
+
+	//for tracking if the heartbeat reporter is running or done running
+	doneCh chan bool
+
+	log logr.Logger
+}
+
+// SetupStatusReporter sets up a singleton of the type `statusReporter` (only if it doesn't exist yet) and returns it to the caller.
+func SetupStatusReporter(addonInstanceInteractor client, addonName string, addonTargetNamespace string, logger logr.Logger) *StatusReporter {
+	initializeSingletonOnce.Do(func() {
+		statusReporterSingleton = &StatusReporter{
+			addonInstanceInteractor: addonInstanceInteractor,
+			addonName:               addonName,
+			addonTargetNamespace:    addonTargetNamespace,
+			latestConditions: []metav1.Condition{
+				{
+					Type:    AddonHealthyConditionType,
+					Status:  metav1.ConditionUnknown,
+					Reason:  "NoHealthReported",
+					Message: fmt.Sprintf("Addon %q hasn't reported health yet", addonName),
+				},
+			},
+			updateCh: make(chan updateOptions),
+			doneCh:   make(chan bool),
+			log:      logger,
+		}
+		// because the heartbeat reporter still hasn't been started
+		defer close(statusReporterSingleton.doneCh)
+	})
+	return statusReporterSingleton
+}
+
+func (sr *StatusReporter) Start(ctx context.Context) error {
+	// ensures to tie only one heartbeat-reporter loop at a time to a StatusReporter object
+	select {
+	case <-sr.doneCh:
+		sr.log.Info("StatusReporter already found to be running.")
+		return nil
```

Comment:
Ahh my bad. Fixing this!
Rather the loop itself should come under this case because a receive on `doneCh` conveys that doneCh was closed hence, it's ok to run the loop. 
This will eradicate the `default` as well.